### PR TITLE
fix: add main thread checker for the app lifecycle integration

### PR DESCRIPTION
**Pull request by marandaneto**
## :loudspeaker: Type of change

- [X] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring

## :scroll: Description
fix: add main thread checker for the app lifecycle integration
this will make the App. to not crash it, but still won't be able to add the observer, see next steps.

## :bulb: Motivation and Context
https://github.com/getsentry/sentry-android/issues/520

## :green_heart: How did you test it?
unit test and running on older androidx versions.

## :pencil: Checklist

- [X] I reviewed submitted code
- [X] I added tests to verify changes
- [X] All tests passing
- [X] No breaking changes

## :crystal_ball: Next steps
maybe execute the addObserver on the main Looper

### CODE REVIEW

File: sentry-core/src/main/java/io/sentry/core/Sentry.java [hunk 1]
```diff
@@ -133,11 +133,10 @@ private static synchronized void init(
...
       options
-        .getLogger()
-        .log(
-          SentryLevel.WARNING,
-          "Sentry has been already initialized. Previous configuration will be overwritten."
-        );
+          .getLogger()
+          .log(
+              SentryLevel.WARNING,
+              "Sentry has been already initialized. Previous configuration will be overwritten.");
```
Comment (marandaneto) [inline sentry-core/src/main/java/io/sentry/core/Sentry.java:139]: just a spotless change

File: sentry-android-core/src/main/java/io/sentry/android/core/AppLifecycleIntegration.java [hunk 2]
```diff
@@ -44,15 +47,15 @@ public void register(final @NotNull IHub hub, final @NotNull SentryOptions optio
...
-
-        options.getLogger().log(SentryLevel.DEBUG, "AppLifecycleIntegration installed.");
+        if (MainThreadChecker.isMainThread()) {
+          addObserver(hub);
+          options.getLogger().log(SentryLevel.DEBUG, "AppLifecycleIntegration installed.");
+        } else {
+          // some versions of the androidx lifecycle-process require this to be executed on the main
+          // thread.
+          final Handler handler = new Handler(Looper.getMainLooper());
+          handler.post(() -> addObserver(hub));
```
Comment (marandaneto) [inline sentry-android-core/src/main/java/io/sentry/android/core/AppLifecycleIntegration.java:57]: that's my suggestion to just not ignore if it's not the main thread.

Comment (bruno-garcia) [reply sentry-android-core/src/main/java/io/sentry/android/core/AppLifecycleIntegration.java:57]: sounds good

File: sentry-android-core/src/test/java/io/sentry/android/core/util/MainThreadCheckerTest.kt [hunk 3]
```diff
@@ -0,0 +1,28 @@
+package io.sentry.android.core.util
+
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import kotlin.test.Test
+import kotlin.test.assertFalse
+import kotlin.test.assertTrue
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+class MainThreadCheckerTest {
```
Comment (philipphofmann) [inline sentry-android-core/src/test/java/io/sentry/android/core/util/MainThreadCheckerTest.kt:10]: `m`: Nice tests, would it make sense to add a test, that starts a thread, calls `MainThreadChecker.isMainThread()` inside it and makes sure it is false?

File: sentry-android-core/src/main/java/io/sentry/android/core/util/MainThreadChecker.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    return isMainThread(Thread.currentThread());
+  }
+
+  /**
+   * Checks if a given thread is the Android Main/UI thread
+   *
+   * @param sentryThread the SentryThread
+   * @return true if it is the main thread or false otherwise
+   */
+  public static boolean isMainThread(final @NotNull SentryThread sentryThread) {
```
Comment (philipphofmann) [inline sentry-android-core/src/main/java/io/sentry/android/core/util/MainThreadChecker.java:39]: `m`: If I am not mistaken, we don't have a test for this.

File: sentry-android-core/src/main/java/io/sentry/android/core/DefaultAndroidEventProcessor.java [hunk 5]
```diff
@@ -181,7 +181,7 @@ private void processNonCachedEvent(final @NotNull SentryEvent event) {
 
     if (event.getThreads() != null) {
       for (SentryThread thread : event.getThreads()) {
-        thread.setCurrent(Looper.getMainLooper().getThread().getId() == thread.getId());
+        thread.setCurrent(MainThreadChecker.isMainThread(thread));
```
Comment (philipphofmann) [inline sentry-android-core/src/main/java/io/sentry/android/core/DefaultAndroidEventProcessor.java:184]: Nice that we also use it here 👏

File: sentry-android-core/src/main/java/io/sentry/android/core/AppLifecycleIntegration.java [hunk 6]
```diff
@@ -44,15 +47,15 @@ public void register(final @NotNull IHub hub, final @NotNull SentryOptions optio
...
-                hub,
-                this.options.getSessionTrackingIntervalMillis(),
-                this.options.isEnableSessionTracking(),
-                this.options.isEnableAppLifecycleBreadcrumbs());
-        ProcessLifecycleOwner.get().getLifecycle().addObserver(watcher);
-
-        options.getLogger().log(SentryLevel.DEBUG, "AppLifecycleIntegration installed.");
+        if (MainThreadChecker.isMainThread()) {
+          addObserver(hub);
+          options.getLogger().log(SentryLevel.DEBUG, "AppLifecycleIntegration installed.");
```
Comment (philipphofmann) [inline sentry-android-core/src/main/java/io/sentry/android/core/AppLifecycleIntegration.java:52]: `l`: don't we also want to have this log message in the `else` below?

File: sentry-android-core/src/main/java/io/sentry/android/core/AppLifecycleIntegration.java [hunk 7]
```diff
@@ -44,15 +47,15 @@ public void register(final @NotNull IHub hub, final @NotNull SentryOptions optio
...
-        watcher =
-            new LifecycleWatcher(
-                hub,
-                this.options.getSessionTrackingIntervalMillis(),
-                this.options.isEnableSessionTracking(),
-                this.options.isEnableAppLifecycleBreadcrumbs());
-        ProcessLifecycleOwner.get().getLifecycle().addObserver(watcher);
-
-        options.getLogger().log(SentryLevel.DEBUG, "AppLifecycleIntegration installed.");
+        if (MainThreadChecker.isMainThread()) {
```
Comment (philipphofmann) [inline sentry-android-core/src/main/java/io/sentry/android/core/AppLifecycleIntegration.java:50]: `l`: Just out of curiosity, why don't we just always do?
```
final Handler handler = new Handler(Looper.getMainLooper());
handler.post(() -> addObserver(hub));
```

**Review by philipphofmann — APPROVED** [review]
LGTM, maybe we can improve a few things though.

File: sentry-android-core/src/test/java/io/sentry/android/core/util/MainThreadCheckerTest.kt (hunk 3 above)
Comment (marandaneto) [reply sentry-android-core/src/test/java/io/sentry/android/core/util/MainThreadCheckerTest.kt:10]: it'd make sense, but as I tested them individually, not sure if it's any useful.

File: sentry-android-core/src/main/java/io/sentry/android/core/util/MainThreadChecker.java (hunk 4 above)
Comment (marandaneto) [reply sentry-android-core/src/main/java/io/sentry/android/core/util/MainThreadChecker.java:39]: yep, I will add it

File: sentry-android-core/src/main/java/io/sentry/android/core/AppLifecycleIntegration.java (hunk 6 above)
Comment (marandaneto) [reply sentry-android-core/src/main/java/io/sentry/android/core/AppLifecycleIntegration.java:52]: good catch, the `else` came later for just getting a review and showing an example and forgot to move things, thx

File: sentry-android-core/src/main/java/io/sentry/android/core/AppLifecycleIntegration.java (hunk 7 above)
Comment (marandaneto) [reply sentry-android-core/src/main/java/io/sentry/android/core/AppLifecycleIntegration.java:50]: if it's already in the main thread, we know that the code will be executed right away.
using a handler, we don't control when this code will be executed (most probably immediately), but let's say that an App. uses `handler/post` heavily, we might be delaying this execution for a bit and things might get out of control, just because of that wdyt?

**Comment by billnbell** [comment]
Will this be a release to npm ?

**Comment by bruno-garcia** [comment]
> Will this be a release to npm ?

Once we release the Android SDK, we bump the [react-native SDK's dependency](https://github.com/getsentry/sentry-react-native)

**Comment by zumaawad** [comment]
When this PR will be released?

File: sentry-android-core/src/main/java/io/sentry/android/core/AppLifecycleIntegration.java (hunk 7 above)
Comment (philipphofmann) [reply sentry-android-core/src/main/java/io/sentry/android/core/AppLifecycleIntegration.java:50]: Make sense for me to keep this little extra complexity for such an edge case. Thanks for explaining.

**Review by philipphofmann — APPROVED** [review]
LGTM

File: sentry-android-core/src/test/java/io/sentry/android/core/util/MainThreadCheckerTest.kt (hunk 3 above)
Comment (marandaneto) [reply sentry-android-core/src/test/java/io/sentry/android/core/util/MainThreadCheckerTest.kt:11]: I added a test for AppLifecycleIntegration which basically does that

**Comment by marandaneto** [comment]
> When this PR will be released?

I need to sync with the RN SDK, but most probably this week (early as possible) as we know this is a blocker for some users.
